cmd/soroban-cli/src/events.rs

```diff
@@ -0,0 +1,192 @@
+    contract_ids: Vec<String>,
+
+    /// A set of (up to 5) topic filters to filter events on
+    #[clap(long, multiple = true)]
+    topics: Vec<String>,
+
+    /// Formatting options for outputted events
+    #[clap(
+        long,
+        possible_values = ["console", "json"],
+        default_value = "console"
+    )]
+    format: String,
+}
+
+#[derive(thiserror::Error, Debug)]
+pub enum Error {
+    #[error("invalid ledger range: low bigger than high ({low} > {high})")]
+    InvalidLedgerRange { low: u32, high: u32 },
+
+    #[error("cannot parse contract ID {contract_id}: {error}")]
+    CannotParseContractId {
+        contract_id: String,
+        error: FromHexError,
+    },
+
+    #[error("too many contracts IDs (max 5)")]
+    TooManyContractIds {},
+
+    #[error("too many topic filters (max 5)")]
+    TooManyTopicFilters {},
+
+    #[error("you must specify either an RPC server or sandbox filepath")]
+    TargetRequired {},
+
+    #[error("sandbox filepath does not exist: {path}")]
+    InvalidSandboxFile { path: String },
+
+    #[error(transparent)]
+    Rpc(#[from] rpc::Error),
+
+    #[error(transparent)]
+    Generic(#[from] Box<dyn std::error::Error>),
+}
+
+impl Cmd {
+    pub async fn run(&self, _matches: &clap::ArgMatches) -> Result<(), Error> {
+        if self.contract_ids.len() > 5 {
+            return Err(Error::TooManyContractIds {});
+        }
+
+        if self.topics.len() > 5 {
+            return Err(Error::TooManyTopicFilters {});
+        }
+
+        if self.start_ledger > self.end_ledger {
+            return Err(Error::InvalidLedgerRange {
+                low: self.start_ledger,
+                high: self.end_ledger,
+            });
+        }
+
+        for raw_contract_id in &self.contract_ids {
+            // We parse the contract IDs to ensure they're the correct format,
+            // but since we'll be passing them as-is to the RPC server anyway,
+            // we disregard the return value.
+            utils::contract_id_from_str(raw_contract_id).map_err(|e| {
+                Error::CannotParseContractId {
+                    contract_id: raw_contract_id.clone(),
+                    error: e,
+                }
+            })?;
+        }
+
+        let mut events: Vec<Event> = Vec::new();
+        if let Some(rpc_url) = self.rpc_url.as_ref() {
+            let client = Client::new(rpc_url);
+            let rpc_event = client.get_events(&self.contract_ids, &self.topics)?;
+            events = rpc_event.events;
+        } else if let Some(path) = self.ledger_file.as_ref() {
+            // TODO: Get events from the sandbox.
+            if !path.exists() {
+                return Err(Error::InvalidSandboxFile {
+                    path: path.to_str().unwrap().to_string(),
+                });
+            }
+        } else {
+            return Err(Error::TargetRequired {});
+        }
+
+        for event in &events {
+            if self.format == "json" {
```

Comment:
❗ We should avoid unwrap in the application if we can since the output is pretty scary to a user. Here's an example of what we do elsewhere for JSON generation failures:
https://github.com/stellar/soroban-tools/blob/e4d3ed4d85181eabf745681b69b039aad028cb8f/cmd/soroban-cli/src/read.rs#L179-L184

Reply:
Hmm, fair enough. In my mind, the RPC server will never return junk, but maybe I give users too much credit :laughing: